feat: add extensible transform API for custom field transformations

OrlovEvgeny opened the pull request:
This enables users to extend compose-go with custom protocols and validation                                                                                logic without forking the project, addressing the need for custom port  protocols like http/https and domain name support

- Add RegisterDefaultValue() functions  
- Add EnhancedPortDefaults as example of extensible port transformation 
- Maintain backward compatibility with existing code

example: 
```go
// EnhancedPortDefaults is an example of an enhanced port defaults function
// that supports additional protocols and domain names in the published field.
// This function can be used as a replacement for the default portDefaults
// by registering it with RegisterDefaultValue.
//
// Example usage:
//   transform.RegisterDefaultValue("services.*.ports.*", transform.EnhancedPortDefaults)
//
// This function supports:
// - Additional protocols: "http", "https", "tcp", "udp"
// - Domain names in the published field (e.g., "example.com:80")
// - All existing functionality of the original portDefaults
func EnhancedPortDefaults(data any, _ tree.Path, _ bool) (any, error) {
	switch v := data.(type) {
	case map[string]any:
		// Set default protocol if not specified
		if _, ok := v["protocol"]; !ok {
			v["protocol"] = "tcp"
		}
		// Set default mode if not specified
		if _, ok := v["mode"]; !ok {
			v["mode"] = "ingress"
		}
		// Additional logic can be added here to handle domain names
		// in the published field or other custom transformations
		return v, nil
	default:
		return data, nil
	}
}
```

Related issue: https://github.com/psviderski/uncloud/issues/44

ndeloof commented on transform/defaults.go:
```diff
@@ -20,7 +20,11 @@ import (
+var DefaultValues = map[tree.Path]TransformFunc{}
+
+// For backward compatibility
+var defaultValues = DefaultValues
```
not needed, as a non-public field there's no backward compatibility to be addressed here

ndeloof commented on transform/defaults.go:
```diff
@@ -30,6 +34,20 @@ func init() {
 }
 
+// RegisterDefaultTransformer registers a custom transformer for the given path pattern
+func RegisterDefaultTransformer(path string, transformer TransformFunc) {
```
`RegisterDefaultTransformer` sounds weird to me. What about `RegisterDefaultValue`

ndeloof commented on transform/defaults.go:
```diff
@@ -30,6 +34,20 @@ func init() {
+}
+
+// GetDefaultTransformers returns a copy of the current default transformers
+func GetDefaultTransformers() map[string]TransformFunc {
```
why have this function while `DefaultValues` is public ?

OrlovEvgeny commented:
Hey @ndeloof 
> Can you please squash your commits, rebase and sign-off, so we can merge ?
Done

OrlovEvgeny commented:
I see,It seems something broke after I rebased main
let me fix it

OrlovEvgeny commented:
@ndeloof fixed

ndeloof commented on transform/canonical.go:
```diff
@@ -20,7 +20,11 @@ import (
+// TransformFunc is a function that can transform data at a specific path
+type TransformFunc func(data any, p tree.Path, ignoreParseError bool) (any, error)
+
+// For backward compatibility
```
unnecessary

OrlovEvgeny commented:
Linter complains about "stuttering" when using type as `transform.TransformFunc`
https://github.com/compose-spec/compose-go/actions/runs/16293608125/job/46012331031#step:4:33

I refactored some naming:
renamed `TransformFunc` to `Func` in transform/canonical.go
what do you think about it?

glours reviewed (commented):
Sounds good!
I wonder if we should not add a test, to avoid breaking change on the feature and give an exemple for developers who want to use it, WDYT?

OrlovEvgeny commented:
> Sounds good! I wonder if we should not add a test, to avoid breaking change on the feature and give an exemple for developers who want to use it, WDYT?

Yes, I think that's a good idea 

I added a test and an example function `exampleEnhancedPortDefaults`

glours reviewed (approved):
LGTM, thanks a lot for this contribution 🙏
